treehasher.go

```diff
@@ -32,9 +32,8 @@ func (th *treeHasher) path(key []byte) []byte {
 }
 
 func (th *treeHasher) digestLeaf(path []byte, leafData []byte) ([]byte, []byte) {
-	value := make([]byte, len(leafPrefix))
-	copy(value, leafPrefix)
-
+	value := make([]byte, 0, len(leafPrefix)+len(path)+len(leafData))
+	value = append(value, leafPrefix...)
```

Comment:
`copy` should be faster than `append`

Reply:
@robert-zaremba It should be the same if we pre-allocate the slice, which we did here.

Reply:
as far as I remember `copy` is faster. Maybe something changed in the recent Go releases.

Reply:
As far as I remember it will literally call the same code under the hood if the slice is pre-allocated. Append is just slightly more idiomatic. Especially, if it is chained / composed of multiple appends. I'm happy to verify this with a benchmark.

Reply:
@robert-zaremba Here's a quick prove:

```
package main

import "testing"

var data = [512]int{}

func BenchmarkAppend(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := make([]int, 0, 512)
		s = append(s, data[:]...)
	}
}

func BenchmarkCopy(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := make([]int, 512)
		copy(s, data[:])
	}
}
```

The result:

```
goos: darwin
goarch: arm64
BenchmarkAppend-8   	 8270780	       136.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkCopy-8     	 8787560	       136.6 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	command-line-arguments	2.793s
```

Reply:
very good. Thanks @cuonglm